### PR TITLE
[Merged by Bors] - TO-2999 add path for db to config

### PR DESCRIPTION
**Pull request by rustonaut**
This does following:

- expose the application data we already have in the configuration in dart to rust
- use that dir for the db (file name `db.sqlite`)
- add another `use_in_memory_db` option to `InitConfig`
  - this option is for integration tests
  - to reduce complexity this option is always there no matter the compiler feature
  

**References:**

- [TO-2999]

[TO-2999]: https://xainag.atlassian.net/browse/TO-2999?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: discovery_engine_core/bindings/src/types/init_config.rs [hunk 1]
```diff
@@ -213,6 +213,28 @@ pub unsafe extern "C" fn init_config_place_of_de_config(
...
+/// Returns a pointer to the `use_in_memory_db` field of a configuration.
+///
+/// # Safety
+///
+/// The pointer must point to a valid [`InitConfig`] memory object,
+/// it might be uninitialized.
+#[no_mangle]
+pub unsafe extern "C" fn init_config_place_of_use_in_memory_db(place: *mut InitConfig) -> *mut u8 {
+    unsafe { addr_of_mut!((*place).use_in_memory_db).cast() }
+}
```
Comment (rustonaut) [inline discovery_engine_core/bindings/src/types/init_config.rs:236]: On some platforms dart ffigen doesn't handle `bool` correctly :shrug: 

This has been somewhat improve in newer versions (I'm not sure if they are released yet) but in consideration that it will soon all be removed it makes little sense to spent additional time and trying to fix it again.

File: discovery_engine_core/core/src/engine.rs [hunk 2]
```diff
@@ -294,6 +294,10 @@ impl Engine {
         let search_config = SearchConfig {
             max_docs_per_batch: config.max_docs_per_search_batch as usize,
         };
+        #[cfg(feature = "storage")]
+        let data_dir = std::path::PathBuf::from(&config.data_dir);
+        #[cfg(feature = "storage")]
+        let use_in_memory_db = config.use_in_memory_db;
```
Comment (rustonaut) [inline discovery_engine_core/core/src/engine.rs:300]: Config is consumed below.

Generally I think there is a lot of potential to cleanup code around `new`, `from_config` the structure of the context etc. but this is outside of the scope of this PR.

File: discovery_engine_core/core/src/engine.rs [hunk 3]
```diff
@@ -1462,6 +1472,8 @@ pub(crate) mod tests {
                     .unwrap_or(u32::MAX),
                 de_config: None,
                 log_file: None,
+                data_dir: "tmp_test_data_dir".into(),
+                use_in_memory_db: true,
```
Comment (rustonaut) [inline discovery_engine_core/core/src/engine.rs:1476]: At some point we probably need to setup some testing with a `TempDir` `data_dir` and proper teardown/setup but that's outside of the scope of this PR but I might need to do it for TY-3061.

File: discovery_engine_core/core/src/engine.rs (hunk 2 above)
Comment (janpetschexain) [reply discovery_engine_core/core/src/engine.rs:301]: i agree, it's also better to finish the db migration first and after the dead code is removed, we can then check what's left to simplify.

File: discovery_engine_core/core/src/config.rs [hunk 4]
```diff
@@ -65,6 +65,10 @@ pub struct InitConfig {
     pub de_config: Option<String>,
     /// Log file path.
     pub log_file: Option<String>,
+    /// Directory in which user data should be stored.
+    pub data_dir: String,
+    /// Use a in-memory db instead of a db in the `data_dir`
+    pub use_in_memory_db: bool,
```
Comment (janpetschexain) [inline discovery_engine_core/core/src/config.rs:71]: ❓ wdyt about merging both fields into an `Option<String>` which either implies in-memory db if `None` or a db directory if `Some(data_dir)`?

Comment (rustonaut) [reply discovery_engine_core/core/src/config.rs:71]: I had considered this, but in the end this is the generic application data directory which is used for more things then just the database.  Using `Option<String>` would work currently as we only use it from rust for the db. But IMHO it would sooner or later brake with us moving more parts to rust. Which is why I kept it separate.

**Review by janpetschexain — APPROVED** [review]

**Comment by rustonaut** [comment]
bors merge
